repo_remote.go

```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+func (r *Repository) RemotesList(opts ...RemotesListOptions) ([]string, error) {
+	return RepoRemotesList(r.path, opts...)
+}
+
+// RemoteURLGetOptions contains arguments for retrieving URL(s) of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteURLGetOptions struct {
+	// False: get fetch URLs
+	// True: get push URLs
+	Push bool
+	// True: get all URLs (lists also non-main URLs; not related with Push)
+	All bool
+	// The timeout duration before giving up for each shell command execution.
+	// The default timeout duration will be used when not supplied.
+	Timeout time.Duration
+}
+
+// RepoRemoteURLGet retrieves URL(s) of a remote of the repository in given path.
+func RepoRemoteURLGet(repoPath, name string, opts ...RemoteURLGetOptions) ([]string, error) {
+	var opt RemoteURLGetOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "get-url")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+	if opt.All {
+		cmd.AddArgs("--all")
+	}
+
+	stdout, err := cmd.AddArgs(name).RunInDirWithTimeout(opt.Timeout, repoPath)
+	if err != nil {
+		return nil, err
+	}
+	return stdoutToStringSlice(stdout), nil
+}
+
+// RemoteURLGet retrieves URL(s) of a remote of the repository in given path.
+func (r *Repository) RemoteURLGet(name string, opts ...RemoteURLGetOptions) ([]string, error) {
+	return RepoRemoteURLGet(r.path, name, opts...)
+}
+
+// RemoteURLSetOptions contains arguments for setting an URL of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteURLSetOptions struct {
+	// False: set fetch URLs
+	// True: set push URLs
+	Push bool
+	// The timeout duration before giving up for each shell command execution.
+	// The default timeout duration will be used when not supplied.
+	Timeout time.Duration
+}
+
+// RepoRemoteURLSetFirst sets first URL of the remote with given name of the repository in given path.
+func RepoRemoteURLSetFirst(repoPath, name, newurl string, opts ...RemoteURLSetOptions) error {
+	var opt RemoteURLSetOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "set-url")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+
+	_, err := cmd.AddArgs(name, newurl).RunInDirWithTimeout(opt.Timeout, repoPath)
+	if err != nil && strings.Contains(err.Error(), "No such remote") {
+		return ErrRemoteNotExist
+	}
+	return err
+}
+
+// RemoteURLSetFirst sets the first URL of the remote with given name of the repository.
+func (r *Repository) RemoteURLSetFirst(name, newurl string, opts ...RemoteURLSetOptions) error {
+	return RepoRemoteURLSetFirst(r.path, name, newurl, opts...)
+}
+
+// RepoRemoteURLSetRegex sets the first URL of the remote with given name and URL regex match of the repository in given path.
+func RepoRemoteURLSetRegex(repoPath, name, urlregex string, newurl string, opts ...RemoteURLSetOptions) error {
```

Comment:
```suggestion
func RepoRemoteSetURL(repoPath, name, urlregex string, newurl string, opts ...RemoteURLSetOptions) error {
```

We should name it like this to follow how we use git bin?

Reply:
regex: I decided in the past to split 'first' and regex for ease of use and clarity.

Alternatively: if `urlregex == ""`, could be a thing, but it implies needing an urlregex param (hope people read the comments). Throwing it under options doesn't sit well with me either.

***
> We should name it like this to follow how we use git bin?
 
In my initial suggestion, I was aiming for hierarchy, and set/get as an operation (`repo/remotes/URLs/set,setregex,get`).

Though, I'm also fine for opting to follow git bin. LMK if you'd like that.

Reply:
> regex: I decided to split 'first' and regex for ease of use and clarity.
> 
> (Alternatively: if `urlregex == ""`, could be a thing, but it implies needing an `urlregex` param (hope people read the docs?). Throwing it under options doesn't sit well with me either.

Ah I see, did not realize this, I guess I'm OK with the "Regex" in the function name then.

> > We should name it like this to follow how we use git bin?
> 
> In my initial suggestion, I was aiming for hierarchy, and set/get as an operation (`repo/remotes/URLs/set,setregex,get`).
> 
> Though, I'm also fine for opting to follow git bin. LMK if you'd like that.

Let's all-in to follow how we use git bin!

Reply:
After opening up the code locally, I think it makes more sense to make the URL regex be part of the options, as it is an option after all.